File: scripts/makefeat.py
#!/usr/bin/env python

# AUTHORS:
# Lilly An (Github: swamplilly)
# Ryan Gurnick

import fileinput, sys, csv
from sys import argv

########################################
#
# Error checking
#
########################################
input_count = len(argv)

if input_count != 3:
    print("ERROR: Missing template file.\nPlease input your template file as follows:\nmakefeat <NAME_OF_TEMPLATE_FILE>")
    sys.exit()

config = argv[1]
template = argv[2]

csv_tag = ".csv"
template_tag = "-template.feature"
feature_tag = ".feature"

if not csv_tag in config:
    print("ERROR: Variables are not in csv format.\nIs variables.csv in the same directory?\nIf you are using another variable file, does it end with .csv?")
    sys.exit()

if not template_tag in template:
    print("ERROR: Input file is not tagged as template.\nPlease make sure your template file is named as: <NAME_OF_FILE>-template.feature")
    sys.exit()



########################################
#
# Declare variables
#
########################################

#create .feature file
new_feature = template.replace(template_tag, feature_tag)
new_feature_fd = open(new_feature, "w+")

#store elements from config
config_fd = open(config, "r")
config_reader = csv.reader(config_fd)
config_dict = {}
for element in config_reader:
    if not element:
        continue
    key = '[%s]' % element[0]
    value = '"%s"' % element[1]
    config_dict[key] = value

#open template file
template_fd = open(template, "rw")
template_lines = [line.rstrip('\n') for line in template_fd]



########################################
#
# Variable substitution
#
########################################

for line in template_lines:
    this_line = line
    for this, with_this in config_dict.iteritems():
        if this in this_line:
            this_line = this_line.replace(this, with_this)
    new_feature_fd.write(this_line + "\n")

File: scripts/makefeat.php
<?php

////////////////////////////////////////////////////////////////////////////////
//
// ERROR CHECKING
//
////////////////////////////////////////////////////////////////////////////////

if (count($argv) != 3) {
    exit("ERROR: Missing feature file or config file.\n* Please execute commands as follows: php maketemp <CONFIG_FILE> <FEATURE_FILE>\n");
}

$config = $argv[1];
$template = $argv[2];

$csv_tag = ".csv";
$template_tag = "-template.feature";
$feature_tag = ".feature";

if (strpos($config, $csv_tag) === false) {
    exit("ERROR: Variables are not in csv format.\n* Is variables.csv in the same directory?\n* If you are using another variable file, does it end with .csv?\n");
}

if (strpos($template, $template_tag) === false) {
    exit("ERROR: Input file is a template.\n* Please make sure your template file is named as: <NAME_OF_FILE>.feature, WITHOUT -template.feature\n");
}

////////////////////////////////////////////////////////////////////////////////
//
// DECLARE VARIABLES
// OPEN FILES
//
////////////////////////////////////////////////////////////////////////////////

// create -template.feature file
$new_feature = str_replace($template_tag, $feature_tag, $template);
$new_feature_fd = fopen($new_feature, "w+");

// store elements from config
$config_dict = array_map('str_getcsv', file($config));

//read feature file
$template_lines = file($template, FILE_IGNORE_NEW_LINES);

////////////////////////////////////////////////////////////////////////////////
//
// VARIABLE SUBSTITUTION
//
////////////////////////////////////////////////////////////////////////////////

foreach($template_lines as $line) {
    $this_line = $line . "\n";
    foreach($config_dict as $def) {
        $key = "[" . $def[0] . "]";
        $val = "\"" . $def[1] . "\"";

        if (strpos($this_line, $key) !== false) {
            $this_line = str_replace($key, $val, $line) . "\n";
        }
    }
    fwrite($new_feature_fd, $this_line);
}

?>